### PR TITLE
Ancestry bug with acts_as_list

### PR DESCRIPTION
**Pull request by fabn**
@swanandp I've made a working solution for my issue described in #131 

I've used ancestry gem to reproduce the issue. The fix make all tests (including mine) passing so it should be good to merge.

Let me know what do you think.

### CODE REVIEW

**Comment by burnzoire** [comment]
any word on this PR getting resolved?

**Comment by brendon** [comment]
That's interesting! I've used ancestry and aal for a long time together and haven't noticed this glitch. I must do some tests to see if it's been happening all along!

I can merge this, but I'd like to get @swanandp's input first.

**Comment by brendon** [comment]
Funny! I must have worked around this in my own way. I don't use a scope on the `acts_as_list` declaration but have the following:

```
  # This helps us build properly ordered lists where ancestry is null (root)
  def acts_as_list_has_ancestry
    if ancestry
      "= #{self.class.connection.quote(ancestry)}"
    else
      'IS NULL'
    end
  end
  
  def scope_condition
    "ancestry #{acts_as_list_has_ancestry} AND decision_tree_owner_id = #{self.class.connection.quote(decision_tree_owner_id)} AND decision_tree_owner_type = #{self.class.connection.quote(decision_tree_owner_type)}"
  end
```

Just out of interest, this is how I've managed to reliably move nodes, and orderly delete sections of a tree:

```
def move(sibling_ids, parent_id)
    remove_from_list

    # find_by_id will return nil for root items rather than raise a nil error
    update_attribute :parent, decision_tree_owner.decisions.find_by_id(parent_id)

    if sibling_ids.first == id.to_s
      if sibling_ids.size == 1
        insert_at
      else
        insert_at decision_tree_owner.decisions.find(sibling_ids.second).position
      end
    else
      insert_at decision_tree_owner.decisions.find(sibling_ids[sibling_ids.index(id.to_s) - 1]).position + 1
    end
  end

  # We need to destroy not in one big lot but child by child and rely on their callbacks to delete further children.
  def apply_orphan_strategy
    unless new_record?
      Decision.unscoped do
        Decision.where(child_conditions).each do |child|
          child.destroy
        end
      end
    end
  end
```

Probably won't help anyone, but you never know.

In terms of the actual fix, hopefully we can get that merged soon.

**Comment by brendon** [comment]
I've confirmed the problem in my own project too. Because I was defining my own `scope_condition` there was no `scope_changed?` method (thus avoiding the bug).

**Comment by swanandp** [comment]
I like the `with_old_attributes` pattern. The only thing I am not sure is about adding ancestry as a dependency for 1 particular test case.  We should probably add a module that can be mixed in when the two gems have to be used together.

**Comment by fabn** [comment]
@swanandp putting the dependency in Gemfile makes it a development dependency.

**Comment by brendon** [comment]
Can the problem be abstracted away from ancestry itself? The issue is with cascading updates isn't it? If that's the case, perhaps the test case could do this in another way?

**Comment by swanandp** [comment]
@fabn Can you try @brendon's suggestion of trying to get a failing test case for what ancestry does, but without a direct dependency on it? With that, I think we can go ahead and merge this in.

**Comment by swanandp** [comment]
Also, master received a few commits, so can you please merge master into your branch?

**Comment by swanandp** [comment]
@fabn Do you want to give this another shot ?

**Comment by fabn** [comment]
@swanandp I'll try to do that in the next few days. There is another pull request that may be related to this. I'll try to rebase on current master to see if the issue is still present.

**Comment by fabn** [comment]
Rebased and pushed, build is still green. Do you still think that having ancestry as development dependency is so bad? If so I'll spend some time as soon as I can to get rid of it, otherwise we can merge this as is. Let me know.

**Comment by fabn** [comment]
@brendon I introduced pluck in fdabc6be149533b3c66367516c3c1cf1223c2b2e, however it works only with Rails 4.0 and above. Currently this gem still targets Rails 3.2 so I should revert that or we must drop 3.2 compatibility at some point. WDYT?

**Comment by brendon** [comment]
Hi @fabn :) Let's not drop 3.2 for now. I still use it in one of my very big apps, and I know many will be using it for quite a while yet :)

Pluck seems to exist in Rails 3.2 in the documentation ...  and I've just checked in the console and it's working, so this is safe to have :)

File: lib/acts_as_list/active_record/acts/list.rb [hunk 1]
```diff
@@ -455,24 +455,31 @@ def update_positions
...
 
           # Temporarily swap changes attributes with current attributes
-          def swap_changed_attributes
-            @changed_attributes.each do |k, _|
-              if self.class.column_names.include? k
-                @changed_attributes[k], self[k] = self[k], @changed_attributes[k]
-              end
+          def with_old_attributes
+            begin
+              changed_attributes.each { |k, _| @changed_attributes[k], self[k] = self[k], @changed_attributes[k] }
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: Just a thought to make this completely use the external interface. Could you not just do:

```ruby
cached_changes = changes

cached_changes.each { |attribute, values| self[attribute] = values[0] }
yield
cached_changes.each { |attribute, values| self[attribute] = values[1] }
```
Then we no longer have to worry about that instance variable.

**Comment by brendon** [comment]
I don't think having ancestry as a development dependency is too bad of an idea now. I know a lot of people use this combination anyway (I do). @swanandp, what do you think?

File: lib/acts_as_list/active_record/acts/list.rb [hunk 2]
```diff
@@ -455,24 +455,31 @@ def update_positions
...
           end
 
           def check_scope
             if internal_scope_changed?
-              swap_changed_attributes
-              send('decrement_positions_on_lower_items') if lower_item
-              swap_changed_attributes
+              with_old_attributes do
+                # if position unchanged and db has a different value for position column
+                unless send("#{position_column}_changed?")
```
Comment (brendon) [inline lib/acts_as_list/active_record/acts/list.rb]: This won't be true anymore if you use my method of swapping attributes unfortunately.

**Comment by swanandp** [comment]
Let's go ahead with ancestry, because at the minimum it gives a nice new test case.

File: lib/acts_as_list/active_record/acts/list.rb (hunk 2 above)
Comment (fabn) [inline lib/acts_as_list/active_record/acts/list.rb]: Yes, that's right.

**Comment by fabn** [comment]
Rebased on latest master.

**Comment by brendon** [comment]
@fabn, I've managed to get the tests to pass in Rails 5: https://github.com/swanandp/acts_as_list/pull/203. I had to implement my idea of attribute swapping as relying on the internal instance variable was breaking tests. What should we do here. Should I merge those Rails 5 changes into master and then you rebase off that?

**Comment by brendon** [comment]
Hi @fabn, I ended up merging the Rails 5 fixes. Can you rebase? There was a bug in the attribute swapper (I don't think @changed_attributes is a think in Rails 5).

**Comment by brendon** [comment]
Hi @fabn, can you rebase this against master and we'll look to get this merged. It'll require a bit of rework as I completely changed the attribute swapping process.

**Comment by swanandp** [comment]
@fabn Feel free to move this from `fabn/acts_as_list:ancestry-bug` to `swanandp/acts_as_list:ancestry-bug`

**Comment by brendon** [comment]
@fabn, a lot has changed now. Do you want to pursue this still? It'd be interesting to see if the problem still exists :)

**Comment by fabn** [comment]
@brendon  I'll try to address this as soon as I have some spare time (very scarce resource...)

Meanwhile I want to thank you for the great job you're doing with this gem.

**Comment by brendon** [comment]
Thanks @fabn :) You're most welcome. It's been great to be experienced enough to contribute back now.

**Comment by alassiter** [comment]
Before I use this, because I use ancestry, what is the status of any conflicts with ancestry?

Thank you

**Comment by brendon** [comment]
Hi @alassiter, I still use `acts_as_list` and `ancestry` together without problems. See toward the end of https://github.com/swanandp/acts_as_list/issues/131 for some methods I find helpful.

If you want to do a bit of work on the idea of this Issue and Pull Request please do. There may still be an issue there but I've not encountered it.

I don't mind bundling ancestry as a test dependency at least in the beginning of trying to figure out if there's still a problem.

**Comment by brendon** [comment]
Hi @fabn, are you still wanting to work on this one? :D

**Comment by brendon** [comment]
Closing this due to inactivity.

**Pull request closed without merging**